### ISSUE TITLE
Compile commands generated for module interfaces in MSVC Preview 4 use unknown command line options

### ISSUE THREAD

**Issue by teschnei** [build2/build2#134]
Compiling a module interface in MSVC 16.9.0 preview 4 generates the following warnings:
```
cl : Command line warning D9002 : ignoring unknown option '/module:interface'
cl : Command line warning D9002 : ignoring unknown option '/module:output'
```

Comparing with the commands that are generated using MSBuild instead, I think that `/module:interface` was renamed to `/interface` and `/module:output` was renamed to `/ifcOutput`.

In addition, compiling a source file that consumes a module interface generates the following warning:
```
cl : Command line warning D9002 : ignoring unknown option '/module:reference'
```
This one appears to have been renamed to `/reference`.    Manually changing the command line options and retrying the commands allow it to compile.

For completeness, here's my compiler options:
```
  cxx        cl@C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.28.29828\bin\HostX64\x64\cl.exe
  mode       /std:c++latest /Zc:__cplusplus /permissive- /D__cpp_modules=201810 /experimental:module
  id         msvc
  version    19.28.29828
  major      19
  minor      28
  patch      29828
  signature  Microsoft (R) C/C++ Optimizing Compiler Version 19.28.29828 for x64
  checksum   a8d3f7d4f55df3e5c41c9f5176b1815ec35b3911874c0f9ca8daa51f3773d251
  target     x86_64-microsoft-win32-msvc14.2
  runtime    msvc
  stdlib     msvcp
  c stdlib   msvc
  ```

I am not sure when the compiler options got renamed since I only started looking at modules in build2 today.

**Comment by Klaim** [comment]
Yeah I already pointed that to the dev team some months ago, the issue is simply that MSVC changed the interface in the end of last year, as part of their solidification of their standard module implementation, but build2 still use the previous interface which is deprecated.
The dev team already know this and what to do, they just didn't have time to update build2 yet (work on gcc was first priority if I understand correctly).

Maybe let's keep this issue as an info source for everybody, I should have done that initially ^^;

**Comment by Klaim** [comment]
Here are the relevant information: https://devblogs.microsoft.com/cppblog/standard-c20-modules-support-with-msvc-in-visual-studio-2019-version-16-8/ see the "Compiler Switch Overhaul" section.

So it have been changed since 16.8 beta, [16.8 have been released in November 2020](https://docs.microsoft.com/en-us/visualstudio/releases/2019/release-notes#--visual-studio-2019-version-1680).

**Comment by boris-kolpackov** [comment]
Yes, we've seen this in our CI also. The gist of it is that `build2` support for modules with MSVC is currently non-functional (both due to this changes but more importantly due to outstanding bugs in MSVC). I am planning to take a look at it in the coming weeks but whether anything can be done to improve the situation remains to be seen.

**Comment by VincentBerthier** [comment]
Does it work on g++ / clang though?

I've just tried to put back the assert for modules with cxx.std=experimental, and I still have the "project/build/root.build:5:1: error: compiler does not support modules" error message.
And that's with g++ (GCC) 11.0.0 20210203 (experimental) [c++-modules revision 20210313-2122] / clang version 12.0.0 (https://github.com/llvm/llvm-project.git 23cc838099e10b13a32e54105f4db0f1b7e3a842) (so both versions updated from their sources as of 2 days ago).

**Comment by Klaim** [comment]
> Does it work on g++ / clang though?
> 
> I've just tried to put back the assert for modules with cxx.std=experimental, and I still have the "project/build/root.build:5:1: error: compiler does not support modules" error message.
> And that's with g++ (GCC) 11.0.0 20210203 (experimental) [c++-modules revision 20210313-2122] / clang version 12.0.0 (https://github.com/llvm/llvm-project.git 23cc838099e10b13a32e54105f4db0f1b7e3a842) (so both versions updated from their sources as of 2 days ago).

Yes for GCC, see details there: https://build2.org/blog/build2-cxx20-modules-gcc.xhtml

Clang: no, because clang does not currently provides C++ modules, and AFAIK it's still not available in v12.0 (or they merged it very recently, but I didn't notice so far)

**Comment by VincentBerthier** [comment]
Ha, that's great news, thanks a lot!

**Comment by boris-kolpackov** [comment]
The MSVC bug that blocks us is still not fixed, so I am re-targeting this to `0.17.0`.

**Comment by boris-kolpackov** [comment]
The upcoming `0.17.0` release has been updated with MSVC modules support (for details, see https://github.com/build2/build2/issues/333) so I am going to close this issue.
